producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,317 @@
+	// Init initializes the consumer writer.
+	Init()
+
+	// Close closes the consumer writer.
+	Close()
+}
+
+type consumerWriterMetrics struct {
+	ackError          tally.Counter
+	decodeError       tally.Counter
+	encodeError       tally.Counter
+	resetConn         tally.Counter
+	resetError        tally.Counter
+	connectError      tally.Counter
+	setKeepAliveError tally.Counter
+}
+
+func newConsumerWriterMetrics(scope tally.Scope) consumerWriterMetrics {
+	return consumerWriterMetrics{
+		ackError:          scope.Counter("ack-error"),
+		decodeError:       scope.Counter("decode-error"),
+		encodeError:       scope.Counter("encode-error"),
+		resetConn:         scope.Counter("reset-conn"),
+		resetError:        scope.Counter("reset-conn-error"),
+		connectError:      scope.Counter("connect-error"),
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type consumerWriterImpl struct {
+	sync.RWMutex
+
+	// encodeLock controls the access to the encode function.
+	encodeLock sync.Mutex
+	// decodeLock controls the access to the decode function.
+	decodeLock sync.Mutex
+	encdec     proto.EncodeDecoder
+
+	addr            string
+	retrier         retry.Retrier
+	router          ackRouter
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	lastResetNanos int64
+	closed         bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	closeWG        sync.WaitGroup
+	m              consumerWriterMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerWriter(
+	addr string,
+	router ackRouter,
+	opts Options,
+) consumerWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	w := &consumerWriterImpl{
+		encdec:          proto.NewEncodeDecoder(defaultConn, opts.EncodeDecoderOptions()),
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		router:          router,
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		lastResetNanos:  0,
+		closed:          false,
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newConsumerWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	w.connectFn = w.connectOnce
+	if err := w.resetWithConnectFn(w.connectFn); err != nil {
+		w.signalReset()
+	}
+	return w
+}
+
+// Write should fail fast so that the write could be tried on other
+// consumer writers that are sharing the data.
+func (w *consumerWriterImpl) Write(m proto.Marshaler) error {
+	w.encodeLock.Lock()
+	err := w.encdec.Encode(m)
+	w.encodeLock.Unlock()
+	if err != nil {
+		w.m.encodeError.Inc(1)
+		w.signalReset()
+	}
+	return err
+}
+
+func (w *consumerWriterImpl) signalReset() {
```

Comment:
I thought about that, but the main issue is that if I can't have a self managed `retryableConnection` because the reset of the connection requires the `encodeLock` and `decodeLock`, for example the main thread might be doing a decoding:

```
w.decodeLock.Lock()
w.decoder.Decode(...)
w.decodeLock.UnLock()
```

meanwhile the `retryableConnection` being used in the decoder might want to swap the connection because it just created a new one, then it also needs to acquire the w.decodeLock to avoid data race during the `Decode` function. So resetting the connection is better to be done on the writer level where it controls all the locks.

Reply:
It maybe be out of scope, but the other option here is to implement a `drain` function on the `proto.EncoderDecoder` object so that you can make you can make sure you have nothing left to decode before you call reset on the `EncoderDecoder` object. You would still need a lock on the decoder but you could have the decoder and connection as separate logical objects.

Reply:
Hmm I don't think we need to drain the old connection here once we have a new connection

Reply:
Why not implement the retryable connection so that encoder and decoder need to acquire a read lock before reading from or writing to the underlying TCP connection, and the retryable connection object needs to acquire a write lock before resetting the connection? For example, something along the lines of
```
type retryableConnection struct {
  sync.RWMutex
  
  conn io.ReadWriteCloser
}

func (c *retryableConnection) Read(p []byte) (int, error) {
  c.RLock()
  n, err := c.conn.Read(p)
  c.RLock()
  return n, err
}

func (c *retryableConnection) Write(p []byte) (int, error) {
  c.RLock()
  n, err := c.conn.Write(p)
  c.RUnlock()
  return n, err
}

func (c *retryableConnection) Reset(conn io.ReadWriteCloser) {
  c.Lock()
  c.conn = conn
  c.Unlock()
}
```

Reply:
Well, the drain here would be to make sure that there are no in-flight decode calls when the connection is swapped. Presumably, if the connection has died there is no data left on the old connection to fetch, but you want to make sure that you are not swapping the connection out while the there is an active decode. 

I am re-reading my comment now and I realize my logic is overly complicated and error-prone. But the idea was that if `retryableConnection` is in its own thread and managing its connection it could detect errors and generate new push new connections to `consumerWriter` when old ones died. The `consumerWriter` could then call `encdec.Drain` to ensure there are no in-flight `Decode` calls and then call reset with the new connection. Again sort of an overly complex pattern but I think it would give you logical separation.

Reply:
@xichen2020 
Hmm, not sure if this works, if the encode received an error and now tries to do a reset, but the decode is blocking on the connection read forever with the `RLock`, then the `Reset` function could never get the `c.Lock`?

@jskelcy 
I don't think `Drain` would work, if you do a `Drain()` and then `Reset()`, how do you make sure there is no in-flight calls between `Drain()` and `Reset()`

Reply:
Encoder and decoder should never directly call reset, instead the connection reset should be managed by the `retryableConnection` object. For example, a connection reset is triggered when there are x number of encode/decode failures and y amount of time has elapsed since last retry.

The connection reset can be done either synchronously or asynchronously. In the latter case, it'll be notification based and might be similar to connection retry logic you have in the `consumerWriterImpl` class.

Reply:
@cw9 yeah that is what I meant by "you would still need a lock".

Reply:
Then the `Reset` function in `retryableConnection` needs to be something like this: 
```
c.conn.Close() // To wake up any blocking encode/decode call
c.Lock()
c.conn = conn
c.Unlock()
```
Which is similar to the current `reset` function in `consumerWriterImpl` as well.

I basically need to move all the connection reset related logic from `consumerWriter` into the `retryableConnection` object and add the lock needed in the `retryableConnection` which is also added to the hot path for both writes/reads. That RLock is purely for resetting the connection because it needs the `encodeLock` and `decodeLock` in `consumerWriter` to guard the accessibility to `write` and `read` functions. Not sure if this worth it performance wise.

Reply:
My suggestion is to not optimize prematurely if such decisions muddy the interface design. Plus, taking a `RLock` is fairly cheap in most cases (fwiw, ingesting a metric on the agg tier or m3db takes quite a few read locks before they are ingested). Currently the connection management logic is very coupled with encoding and decoding logic, which makes it difficult to understand (at least IMO). By decoupling the two you end up with a much better separation of responsibility and encapsulation.

Reply:
Hm what state does encoder/decoder maintain? ideally they should just be stateless.

Reply:
`encoder` holds a bufio.Writer on top of the conn, `decoder` holds a bufio.Reader on top of the conn

Reply:
I guess I could move the `bufio.Writer`/`bufio.Reader` part out to the `retryableConnection` and do the reset to further decouple the encdec and connection

Reply:
Yeah those can be moved inside the retryable connection implementation, and you can still use those in the `Read` and `Write` methods for the retryable connection object I'd think.